web/src/settings_user_topics.ts

```diff
@@ -0,0 +1,111 @@
+import $ from "jquery";
+import assert from "minimalistic-assert";
+
+import render_user_topic_ui_row from "../templates/user_topic_ui_row.hbs";
+
+import * as ListWidget from "./list_widget";
+import * as scroll_util from "./scroll_util";
+import * as settings_config from "./settings_config";
+import * as user_topics from "./user_topics";
+import type {UserTopic} from "./user_topics";
+
+export let loaded = false;
+
+type UserTopicWithStream = UserTopic & {stream: string};
+
+export function populate_list(): void {
+    const all_user_topics = [];
+    const visibility_policies = Object.values(user_topics.all_visibility_policies);
+    for (const visibility_policy of visibility_policies) {
+        /* For visibility_policy=INHERIT, 'get_user_topics_for_visibility_policy' returns
+        an empty list as we only maintain the record of topics having visibility_policy
+        other than INHERIT; INHERIT is used to remove topics from the record. */
+        const user_topics_list =
+            user_topics.get_user_topics_for_visibility_policy(visibility_policy);
+        all_user_topics.push(...user_topics_list);
+    }
+    const $user_topics_table = $("#user_topics_table");
+    const $search_input = $<HTMLInputElement>("#user_topics_search");
+
+    // This view displays stream names, so we need to make sure that all user_topics
+    // have valid stream names.
```

Comment:
This was needed for some typechecking below to work (for the `sort_fields`). Given that this view seems to not have been broken in the past, I think it's safe to assert that the stream is always defined, but maybe it's better to be more robust and default to the empty string?

Reply:
I think it might be possible to have a `UserTopic` for a stream that's deleted or otherwise you can longer access. I'm not sure those can end up at the client via the API ... but probably the correct thing to do is display the translation of "Unknown stream" for such streams.

Reply:
OK looking at this data structure, it certainly looks like values don't get put into these data structures unless they have a stream name:

```
export function set_user_topic(user_topic: ServerUserTopic): void {                                     
    const stream_id = user_topic.stream_id;                                                             
    const topic = user_topic.topic_name;                                                                
    const date_updated = user_topic.last_updated;                                                       
                                                                                                        
    const stream_name = sub_store.maybe_get_stream_name(stream_id);                                     
                                                                                                        
    if (!stream_name) {                                                                                 
        blueslip.warn("Unknown stream ID in set_user_topic: " + stream_id);                             
        return;                                                                                         
    }                                                                                                   
                                                                                                        
    update_user_topics(stream_id, topic, user_topic.visibility_policy, date_updated);                   
}                                                                                                       
```

Reply:
So I think we don't need this code, we just need to fix the `UserTopic` type to disallow `undefined`; which includes probably making `set_user_topic` pass the stream name into `update_user_topics` rather than looking it up again.

Reply:
This particular instance is calling `user_topics.get_user_topics_for_visibility_policy` which gets the stream like this `const stream = sub_store.maybe_get_stream_name(stream_id);` -- but it does only use `stream_id` found in `all_user_topics`.

And `all_user_topics.set` only happens in `update_user_topics`, which aha is where you ended up. Okay great. Yeah I'll update that!

Reply:
Updated :)